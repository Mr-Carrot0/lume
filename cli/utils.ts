import lume from "../mod.ts";
import { log } from "../core/utils/log.ts";
import Site from "../core/site.ts";

/** Create a site instance */
export async function createSite(_config?: URL): Promise<Site> {
  if (!_config) {
    return lume();
  }
  log.info(`Loading config file <gray>${_config}</gray>`);
  const mod = await import(_config.toString());
  const site = mod.default as Site | undefined;
  if (!(site instanceof Site)) {
    log.fatal(
      `[Lume] Missing Site instance! Ensure your config file does export the Site instance as default.`,
    );
    throw new Error("Site instance is not found");
  }
  return site;
}

/** Create a site intance and build it */
export async function buildSite(site: Site): Promise<void> {
  performance.mark("start");
  await site.build();
  performance.mark("end");

  log.info(`üçæ Site built into <gray>${site.options.dest}</gray>`);
  const duration = performance.measure("duration", "start", "end").duration /
    1000;
  const total = site.pages.length + site.files.length;
  log.info(
    `  <gray>${total} files generated in ${duration.toFixed(2)} seconds</gray>`,
  );
}
