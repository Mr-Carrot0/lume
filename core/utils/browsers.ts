/**
 * Code generated from:
 *
 * ```ts
 * import { getCompatibleVersions } from "npm:baseline-browser-mapping@2.4.4";
 *
 * const results = getCompatibleVersions({
 *   widelyAvailableOnDate: "2025-05-07", // Lume 3.0.0 release date
 * });
 * ```
 *
 * Not run directly to avoid importing this package that requires read access.
 */
export const browsers = {
  chrome: [107, 0],
  chrome_android: [107, 0],
  edge: [107, 0],
  firefox: [104, 0],
  firefox_android: [104, 0],
  safari: [16, 0],
  safari_ios: [16, 0],
} as Record<string, [number, number]>;

export function version(browser: [number, number]): number {
  return (browser[0] << 16) | (browser[1] << 8);
}
export function versionString(browser: [number, number]): string {
  return `${browser[0]}.${browser[1]}`;
}
