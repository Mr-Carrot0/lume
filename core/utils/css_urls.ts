type UrlWalker = (url: string, type: string) => string | Promise<string>;
type Position = [string, string, number, number];
type Index = [string, number];

export async function walkUrls(
  code: string,
  walker: UrlWalker,
): Promise<string> {
  let result = "";
  let index = 0;
  let position = next(code, index);

  while (position) {
    const [type, url, start, end] = position;
    const newUrl = await walker(url, type);
    const quotedUrl = url.startsWith("data:")
      ? `'${newUrl.replaceAll("'", "\\'")}'`
      : `"${newUrl.replaceAll('"', '\\"')}"`;
    result += `${code.slice(index, start)}${quotedUrl}`;
    index = end;
    position = next(code, index);
  }

  return result + code.slice(index);
}

function next(code: string, index = 0): Position | undefined {
  const nextUrl = code.indexOf("url(", index);
  const nextComment = code.indexOf("/*", index);
  const nextImport = code.indexOf("@import ", index);

  let found: Index | undefined;

  if (nextUrl !== -1) {
    found = ["url", nextUrl];
  }
  if (nextComment !== -1 && (!found || nextComment < found[1])) {
    found = ["comment", nextComment];
  }
  if (nextImport !== -1 && (!found || nextImport < found[1])) {
    found = ["import", nextImport];
  }

  if (!found) {
    return;
  }

  // If it's a comment, find the end of it
  if (found[0] === "comment") {
    const end = code.indexOf("*/", found[1]);
    if (end === -1) {
      return;
    }
    return next(code, end + 2);
  }

  if (found[0] === "import") {
    const double = code.indexOf('"', found[1]);
    const single = code.indexOf("'", found[1]);
    const urlFn = code.indexOf("url(", found[1]);

    let foundImport: Index | undefined;

    if (double !== -1) {
      foundImport = ["double", double];
    }
    if (single !== -1 && (!foundImport || single < foundImport[1])) {
      foundImport = ["single", single];
    }

    // It's has a url(), so delegate to the url() logic
    if (urlFn !== -1 && (!foundImport || urlFn < foundImport[1])) {
      return next(code, urlFn);
    }

    if (!foundImport) {
      return;
    }

    const start = foundImport[1];
    const end =
      code.indexOf(foundImport[0] === "double" ? '"' : "'", start + 1) + 1;
    return ["url", code.slice(start + 1, end - 1), start, end];
  }

  if (found[0] === "url") {
    const start = code.indexOf("(", found[1]) + 1;
    let url = code.slice(start).trim();
    const quote = url.startsWith('"') || url.startsWith("'")
      ? url[0]
      : undefined;
    url = quote ? url.slice(1) : url;
    let end = findUrlEnd(url, quote);
    url = url.slice(0, end);
    end += start + (quote ? 2 : 0);
    url = url.replace(/\\'/g, "'").replace(/\\"/g, '"');
    return ["url", url, start, end];
  }
}

function findUrlEnd(code: string, quote?: string): number {
  if (!quote) {
    return code.indexOf(")");
  }

  let end = code.indexOf(quote);

  while (end !== -1 && code[end - 1] === "\\") {
    end = code.indexOf(quote, end + 1);
  }

  return end === -1 ? code.indexOf(")") : end;
}
