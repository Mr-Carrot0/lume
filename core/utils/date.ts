/**
 * Returns the date of the git commit that created or modified the file.
 * Thanks to https://github.com/11ty/eleventy/blob/8dd2a1012de92c5ee1eab7c37e6bf1b36183927e/src/Util/DateGitLastUpdated.js
 */
export function getGitDate(
  type: "created" | "modified",
  file: string,
): Date | undefined {
  const args = type === "created"
    ? ["log", "--diff-filter=A", "--follow", "-1", "--format=%at", "--", file]
    : ["log", "-1", "--format=%at", "--", file];

  const { stdout, success } = new Deno.Command("git", { args }).outputSync();

  if (!success) {
    return;
  }
  const str = new TextDecoder().decode(stdout);

  if (str) {
    return parseDate(parseInt(str) * 1000);
  }
}

/** Parse a string or number (of miliseconds) to UTC Date */
export function parseDate(date: string | number): Date {
  return new Date(getZonedDateTime(date).epochMilliseconds);
}

/** Parse a string or number (of miliseconds) to a zoned datetime */
export function getZonedDateTime(
  date: string | number,
  timezone = "UTC",
): Temporal.ZonedDateTime {
  if (typeof date === "number") {
    return Temporal.Instant.fromEpochMilliseconds(date).toZonedDateTimeISO(
      timezone,
    );
  }

  try {
    return Temporal.Instant.from(date).toZonedDateTimeISO(timezone);
  } catch {
    return Temporal.PlainDateTime.from(date).toZonedDateTime(timezone);
  }
}
