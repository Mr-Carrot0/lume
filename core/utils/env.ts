const envVars = new Map<string, string>();

export function setEnv(name: string, value: string) {
  if (allowedEnvVars()) {
    Deno.env.set(name, value);
  }
  envVars.set(name, value);
}

export function env<T>(name: string): T | undefined {
  const value = allowedEnvVars()
    ? envVars.get(name) ?? Deno.env.get(name)
    : envVars.get(name);

  if (typeof value === "undefined") {
    return undefined;
  }

  switch (value.toLowerCase()) {
    case "true":
    case "on":
    case "1":
      return true as T;

    case "false":
    case "off":
    case "0":
      return false as T;

    default:
      return value as T;
  }
}

let allowed: boolean | undefined;
function allowedEnvVars(): boolean {
  if (allowed === undefined) {
    allowed = Deno.permissions.querySync?.({ name: "env" }).state === "granted";
  }
  return allowed;
}
