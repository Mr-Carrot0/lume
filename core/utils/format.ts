const byteUnit = Intl.NumberFormat("en", {
  notation: "compact",
  style: "unit",
  unit: "byte",
  unitDisplay: "narrow",
});

export function bytes(bytes: number) {
  return byteUnit.format(bytes);
}

const percentageUnit = Intl.NumberFormat("en", {
  style: "percent",
  minimumFractionDigits: 1,
  maximumFractionDigits: 1,
});

export function percentage(initialValue: number, finalValue: number) {
  if (initialValue === finalValue) {
    return "0%";
  }

  const ratio = -(1 - finalValue / initialValue);

  if (ratio < 0) {
    return percentageUnit.format(ratio);
  }

  return "+" + percentageUnit.format(ratio);
}

const secondUnit = Intl.NumberFormat("en", {
  notation: "compact",
  style: "unit",
  unit: "second",
  unitDisplay: "short",
  minimumFractionDigits: 1,
  maximumFractionDigits: 1,
});

const milisecondUnit = Intl.NumberFormat("en", {
  notation: "compact",
  style: "unit",
  unit: "millisecond",
  unitDisplay: "short",
  maximumFractionDigits: 0,
});

export function duration(miliseconds: number) {
  if (miliseconds < 1000) {
    return milisecondUnit.format(miliseconds);
  }

  return secondUnit.format(miliseconds / 1000);
}
