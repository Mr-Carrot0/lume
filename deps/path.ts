export * from "jsr:@std/path@1.1.0";
export * as posix from "jsr:@std/path@1.1.0/posix";
