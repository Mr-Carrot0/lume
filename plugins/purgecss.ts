import { PurgeCSS, purgeHtml } from "../deps/purgecss.ts";
import { merge } from "../core/utils/object.ts";
import { concurrent } from "../core/utils/concurrent.ts";
import { getExtension, matchExtension } from "../core/utils/path.ts";

import type Site from "../core/site.ts";
import type { Page } from "../core/file.ts";
import type { RawContent, UserDefinedOptions } from "../deps/purgecss.ts";

export interface Options {
  /** The list of page extensions loaded as content. */
  contentExtensions?: string[];

  /** Options for purgecss */
  options?: Partial<UserDefinedOptions>;
}

// Default options
export const defaults: Options = {
  contentExtensions: [".html", ".js"],
  options: {
    extractors: [],
  },
};

/**
 * A plugin to remove unused CSS
 */
export function purgecss(userOptions?: Options) {
  const options = merge(defaults, userOptions);

  options.options.extractors!.push({
    extractor: purgeHtml,
    extensions: ["html"],
  });

  return (site: Site) => {
    site.process([".css"], async (pages, allPages) => {
      const content: RawContent[] = [];
      for (const page of allPages) {
        if (!matchExtension(options.contentExtensions, page.outputPath)) {
          continue;
        }

        const pageContent = page.content;
        if (typeof pageContent !== "string") {
          return;
        }

        content.push({
          raw: pageContent,
          extension: getExtension(page.outputPath).slice(1),
        });
      }

      await concurrent(pages, async (page: Page) => {
        const pageContent = page.content;
        if (typeof pageContent !== "string") {
          return;
        }

        const purgeOptions: UserDefinedOptions = {
          ...options.options,
          content: (options.options.content || []).concat(content),
          css: [
            {
              raw: pageContent,
            },
          ],
        };

        const result = await new PurgeCSS().purge(purgeOptions);

        page.content = result[0].css;
      });
    });
  };
}

export default purgecss;
