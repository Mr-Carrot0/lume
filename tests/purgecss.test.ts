import { assertSiteSnapshot, build, getPath, getSite } from "./utils.ts";
import { normalizePath } from "../core/utils/path.ts";
import purgecss from "../plugins/purgecss.ts";

Deno.test("purgecss plugin", async (t) => {
  const site = getSite({
    src: "purgecss",
  });

  site.loadAssets([".css", ".js"]);
  site.add("static", ".");

  site.use(purgecss());

  await build(site);
  await assertSiteSnapshot(t, site);
});

Deno.test("purgecss plugin with options", async (t) => {
  const site = getSite({
    src: "purgecss",
  });

  site.loadAssets([".css", ".js"]);
  site.add("static", ".");

  site.use(purgecss({
    options: {
      content: [
        normalizePath(getPath("./assets/purgecss/static/**/*.html")),
        {
          raw: '<img id="img-option" src="">',
          extension: "html",
        },
      ],
      safelist: ["unused"],
      blocklist: ["strong"],
      variables: true,
    },
  }));

  await build(site);
  await assertSiteSnapshot(t, site);
});
