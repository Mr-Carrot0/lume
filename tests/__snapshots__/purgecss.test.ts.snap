export const snapshot = {};

snapshot[`purgecss plugin 1`] = `
{
  formats: [
    {
      engines: 0,
      ext: ".page.toml",
      isPage: true,
      loader: [AsyncFunction: toml],
    },
    {
      engines: 1,
      ext: ".page.ts",
      isPage: true,
      loader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".page.js",
      isPage: true,
      loader: [AsyncFunction: module],
    },
    {
      engines: 0,
      ext: ".page.jsonc",
      isPage: true,
      loader: [AsyncFunction: json],
    },
    {
      engines: 0,
      ext: ".page.json",
      isPage: true,
      loader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".json",
      loader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".jsonc",
      loader: [AsyncFunction: json],
    },
    {
      engines: 1,
      ext: ".md",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      engines: 1,
      ext: ".markdown",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".js",
      loader: [AsyncFunction: module],
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".ts",
      loader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".vento",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      engines: 1,
      ext: ".vto",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      dataLoader: [AsyncFunction: toml],
      engines: 0,
      ext: ".toml",
      loader: [AsyncFunction: toml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yaml",
      isPage: true,
      loader: [AsyncFunction: yaml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yml",
      isPage: true,
      loader: [AsyncFunction: yaml],
    },
    {
      ext: ".css",
    },
  ],
  src: [
    "/",
    "/_includes",
    "/_includes/footer.vto",
    "/_includes/layout.vto",
    "/index.vto",
    "/pages",
    "/pages/page1.md",
    "/pages/page2.page.js",
    "/script.js",
    "/static",
    "/static/static.html",
    "/styles.css",
  ],
}
`;

snapshot[`purgecss plugin 2`] = `
[
  {
    entry: "/script.js",
    flags: [],
    outputPath: "/script.js",
  },
  {
    entry: "/static/static.html",
    flags: [],
    outputPath: "/static.html",
  },
]
`;

snapshot[`purgecss plugin 3`] = `
[
  {
    content: '<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>TÍTULO</title>
</head>
<body>
  <h1>Título</h1>

<strong>strong</strong>

<div class="content-vento" tabindex="0">Content</div>


  <footer>
Título
</footer>
</body>
</html>
',
    data: {
      basename: "index",
      children: '<h1>Título</h1>

<strong>strong</strong>

<div class="content-vento" tabindex="0">Content</div>
',
      content: '<h1>{{ title }}</h1>

<strong>strong</strong>

<div class="content-vento" tabindex="0">Content</div>
',
      date: [],
      layout: "layout.vto",
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      title: "Título",
      url: "/",
    },
    src: {
      asset: false,
      ext: ".vto",
      path: "/index",
      remote: undefined,
    },
  },
  {
    content: '<div class="content-dynamic">Content of Page 2</div>',
    data: {
      basename: "page2",
      children: '<div class="content-dynamic">Content of Page 2</div>',
      content: '<div class="content-dynamic">Content of Page 2</div>',
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      title: "Page 2",
      url: "/page_2/",
    },
    src: {
      asset: false,
      ext: ".page.js",
      path: "/pages/page2",
      remote: undefined,
    },
  },
  {
    content: "<p>Content of Page 1</p>
",
    data: {
      basename: "page1",
      children: "<p>Content of Page 1</p>
",
      content: "Content of Page 1
",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      title: "Page 1",
      url: "/pages/page1/",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/pages/page1",
      remote: undefined,
    },
  },
  {
    content: "Uint8Array(732)",
    data: {
      basename: "styles",
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/styles.css",
    },
    src: {
      asset: true,
      ext: ".css",
      path: "/styles",
      remote: undefined,
    },
  },
]
`;

snapshot[`purgecss plugin with options 1`] = `
{
  formats: [
    {
      engines: 0,
      ext: ".page.toml",
      isPage: true,
      loader: [AsyncFunction: toml],
    },
    {
      engines: 1,
      ext: ".page.ts",
      isPage: true,
      loader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".page.js",
      isPage: true,
      loader: [AsyncFunction: module],
    },
    {
      engines: 0,
      ext: ".page.jsonc",
      isPage: true,
      loader: [AsyncFunction: json],
    },
    {
      engines: 0,
      ext: ".page.json",
      isPage: true,
      loader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".json",
      loader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".jsonc",
      loader: [AsyncFunction: json],
    },
    {
      engines: 1,
      ext: ".md",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      engines: 1,
      ext: ".markdown",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".js",
      loader: [AsyncFunction: module],
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".ts",
      loader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".vento",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      engines: 1,
      ext: ".vto",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      dataLoader: [AsyncFunction: toml],
      engines: 0,
      ext: ".toml",
      loader: [AsyncFunction: toml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yaml",
      isPage: true,
      loader: [AsyncFunction: yaml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yml",
      isPage: true,
      loader: [AsyncFunction: yaml],
    },
    {
      ext: ".css",
    },
  ],
  src: [
    "/",
    "/_includes",
    "/_includes/footer.vto",
    "/_includes/layout.vto",
    "/index.vto",
    "/pages",
    "/pages/page1.md",
    "/pages/page2.page.js",
    "/script.js",
    "/static",
    "/static/static.html",
    "/styles.css",
  ],
}
`;

snapshot[`purgecss plugin with options 2`] = `
[
  {
    entry: "/script.js",
    flags: [],
    outputPath: "/script.js",
  },
  {
    entry: "/static/static.html",
    flags: [],
    outputPath: "/static.html",
  },
]
`;

snapshot[`purgecss plugin with options 3`] = `
[
  {
    content: '<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>TÍTULO</title>
</head>
<body>
  <h1>Título</h1>

<strong>strong</strong>

<div class="content-vento" tabindex="0">Content</div>


  <footer>
Título
</footer>
</body>
</html>
',
    data: {
      basename: "index",
      children: '<h1>Título</h1>

<strong>strong</strong>

<div class="content-vento" tabindex="0">Content</div>
',
      content: '<h1>{{ title }}</h1>

<strong>strong</strong>

<div class="content-vento" tabindex="0">Content</div>
',
      date: [],
      layout: "layout.vto",
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      title: "Título",
      url: "/",
    },
    src: {
      asset: false,
      ext: ".vto",
      path: "/index",
      remote: undefined,
    },
  },
  {
    content: '<div class="content-dynamic">Content of Page 2</div>',
    data: {
      basename: "page2",
      children: '<div class="content-dynamic">Content of Page 2</div>',
      content: '<div class="content-dynamic">Content of Page 2</div>',
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      title: "Page 2",
      url: "/page_2/",
    },
    src: {
      asset: false,
      ext: ".page.js",
      path: "/pages/page2",
      remote: undefined,
    },
  },
  {
    content: "<p>Content of Page 1</p>
",
    data: {
      basename: "page1",
      children: "<p>Content of Page 1</p>
",
      content: "Content of Page 1
",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      title: "Page 1",
      url: "/pages/page1/",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/pages/page1",
      remote: undefined,
    },
  },
  {
    content: "Uint8Array(732)",
    data: {
      basename: "styles",
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/styles.css",
    },
    src: {
      asset: true,
      ext: ".css",
      path: "/styles",
      remote: undefined,
    },
  },
]
`;
